source/server/connection_handler_impl.cc

```diff
@@ -293,54 +307,92 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
+    active_connection->connection_->close(Network::ConnectionCloseType::NoFlush);
   }
 
-  onNewConnection(std::move(new_connection));
-}
-
-void ConnectionHandlerImpl::ActiveTcpListener::onNewConnection(
-    Network::ConnectionPtr&& new_connection) {
-  ENVOY_CONN_LOG(debug, "new connection", *new_connection);
-
   // If the connection is already closed, we can just let this connection immediately die.
-  if (new_connection->state() != Network::Connection::State::Closed) {
-    ActiveConnectionPtr active_connection(
-        new ActiveConnection(*this, std::move(new_connection), parent_.dispatcher_.timeSource()));
+  if (active_connection->connection_->state() != Network::Connection::State::Closed) {
+    ENVOY_CONN_LOG(debug, "new connection", *active_connection->connection_);
+    active_connection->connection_->addConnectionCallbacks(*active_connection);
     active_connection->moveIntoList(std::move(active_connection), connections_);
-    parent_.num_connections_++;
   }
 }
 
-ConnectionHandlerImpl::ActiveConnection::ActiveConnection(ActiveTcpListener& listener,
-                                                          Network::ConnectionPtr&& new_connection,
-                                                          TimeSource& time_source)
+namespace {
+// Structure used to allow a unique_ptr to be captured in a posted lambda. See below.
+struct RebalancedSocket {
+  Network::ConnectionSocketPtr socket;
+};
+using RebalancedSocketSharedPtr = std::shared_ptr<RebalancedSocket>;
+} // namespace
+
+void ConnectionHandlerImpl::ActiveTcpListener::post(Network::ConnectionSocketPtr&& socket) {
+  // It is not possible to capture a unique_ptr so we must bundle the socket inside a shared_ptr
+  // that can be captured.
+  RebalancedSocketSharedPtr socket_to_rebalance = std::make_shared<RebalancedSocket>();
+  socket_to_rebalance->socket = std::move(socket);
+
+  parent_.dispatcher_.post([socket_to_rebalance, tag = config_.listenerTag(), &parent = parent_]() {
+    // TODO(mattklein123): We should probably use a hash table here to lookup the tag instead of
+    // iterating through the listener list.
+    auto listener_it = std::find_if(
+        parent.listeners_.begin(), parent.listeners_.end(),
+        [tag](const std::pair<Network::Address::InstanceConstSharedPtr, ActiveListenerPtr>& p) {
+          return p.second->listener() != nullptr && p.second->listenerTag() == tag;
+        });
+
+    if (listener_it == parent.listeners_.end()) {
+      // It is possible for a listener to be removed while rebalancing.
+      return;
+    }
+
+    // TODO(mattklein123): Remove dynamic_cast here in follow up cleanup.
+    ActiveTcpListener& listener = *dynamic_cast<ActiveTcpListener*>(listener_it->second.get());
+    listener.onAcceptWorker(std::move(socket_to_rebalance->socket),
+                            listener.config_.handOffRestoredDestinationConnections(), true);
```

Comment:
I suspect this would be fewer lines & easier to debug with a for-loop rather than find_if. Something like
```
for (const auto& p : parent_listeners_) {
  if (p.second->listener() != nullptr && p.second->listenerTag() == tag) {
    // TODO(mattklein123): Remove dynamic_cast here in follow up cleanup.
    ActiveTcpListener& listener = *dynamic_cast<ActiveTcpListener*>(listener_it->second.get());
    listener.onAcceptWorker(std::move(socket_to_rebalance->socket),
                            listener.config_.handOffRestoredDestinationConnections(), true);
    break;
  }
}
```